% Version 3 December 2023
% See section 11 of the User Manual for version history
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% \documentclass[referee,sn-basic]{sn-jnl}% referee option is meant for double line spacing

%%=======================================================%%
%% to print line numbers in the margin use lineno option %%
%%=======================================================%%

% \documentclass[lineno,sn-basic]{sn-jnl}% Basic Springer Nature Reference Style/Chemistry Reference Style

%%======================================================%%
%% to compile with pdflatex/xelatex use pdflatex option %%
%%======================================================%%

%%\documentclass[pdflatex,sn-basic]{sn-jnl}% Basic Springer Nature Reference Style/Chemistry Reference Style

%%Note: the following reference styles support Namedate and Numbered referencing. By default the style follows the most common style. To switch between the options you can add or remove Numbered in the optional parenthesis.\\

%%The option is available for: sn-basic.bst, sn-vancouver.bst, sn-chicago.bst%

\documentclass[pdflatex,sn-nature, lineno]{sn-jnl}% Style for submissions to Nature Portfolio journals
% \documentclass[pdflatex,sn-basic]{sn-jnl}% Basic Springer Nature Reference Style/Chemistry Reference Style
% \documentclass[pdflatex, sn-mathphys-num, lineno]{sn-jnl}% Math and Physical Sciences Numbered Reference Style
%%\documentclass[pdflatex,sn-mathphys-ay]{sn-jnl}% Math and Physical Sciences Author Year Reference Style
%%\documentclass[pdflatex,sn-aps]{sn-jnl}% American Physical Society (APS) Reference Style
%%\documentclass[pdflatex,sn-vancouver,Numbered]{sn-jnl}% Vancouver Reference Style
%%\documentclass[pdflatex,sn-apa]{sn-jnl}% APA Reference Style
%%\documentclass[pdflatex,sn-chicago]{sn-jnl}% Chicago-based Humanities Reference Style

%%%% Standard Packages
%% additional latex packages if required can be included here>
\usepackage{graphicx}%
\usepackage{multirow}%
\usepackage{amsmath,amssymb,amsfonts}%
\usepackage{amsthm}%
\usepackage{mathrsfs}%
\usepackage[title]{appendix}%
\usepackage{xcolor}%
\usepackage{textcomp}%
\usepackage{manyfoot}%
\usepackage{booktabs}%
\usepackage{algorithm}%
\usepackage{algorithmicx}%
\usepackage{algpseudocode}%
\usepackage{listings}%

\usepackage{xspace}
\usepackage{standalone}

%% my package start

% for collabration and will remove before submit
\newcommand{\yy}[1]{\textcolor{cyan}{\textbf{[Yangyang: #1]}}}
\newcommand{\rd}[1]{\textcolor{green}{\textbf{[Dr.Yang: #1]}}}
\newcommand{\ty}[1]{\textcolor{orange}{\textbf{[Tingyou: #1]}}}
\newcommand{\qx}[1]{\textcolor{purple}{\textbf{[Qingxiang: #1]}}}

% for collabration and will remove before submit

\usepackage[detect-all=true]{siunitx}
\sisetup{
    scientific-notation=false,
    round-mode = places,
    round-precision = 0
}

\usepackage[acronym, automake, style=index, shortcuts]{glossaries-extra}
\setabbreviationstyle[acronym]{long-short}
% define glossaries
\makeglossaries

\newacronym{blat}{BLAT}{BLAST-like alignment tool}
\newacronym{llm}{LLM}{Large Language Model}
\newacronym{hmm}{HMM}{Hidden Markov Model}
\newacronym{gpu}{GPU}{Graphics Processing Unit}
\newacronym{hpc}{HPC}{High Performance Computing}
\newacronym{bert}{BERT}{Bidirectional Encoder Representations from Transformers}
\newacronym{gpt}{GPT}{Generative Pre-trained Transformer}

\newacronym{ide}{IDE}{Integrated Development Environment}
\newacronym{cd}{CD}{Continuous Development}
\newacronym{ucsc}{UCSC}{UCSC Genome Browser}
\newacronym{glm}{GLM}{Genomic Language Model}
\newacronym{lcglm}{LCGLM}{long-context genomic language model}
\newacronym{snp}{SNP}{Single Nucleotide Polymorphism}
\newacronym{fm}{FM}{Foundational Model}
\newacronym{nlp}{NLP}{Natural Language Processing}

\newacronym{mlp}{MLP}{multilayer perceptron}
\newacronym{drs}{dRNA-seq}{direct RNA sequencing}
\newacronym{ont}{ONT}{Oxford Nanopore Technologies}
\newacronym{pb}{PacBio}{Pacific Biosciences}

\newacronym{go}{GO}{Gene Ontology}
\newacronym{r2c2}{R2C2}{Rolling Circle Amplification to Concatemeric Consensus}
\newacronym{pcr}{PCR}{Polymerase Chain Reaction}

\newacronym{fsm}{FMS}{Full splice match}
\newacronym{ism}{ISM}{Incomplete splice match}
\newacronym{nic}{NIC}{Novel in catalog}
\newacronym{nnc}{NNC}{Novel not in catalog}

\newacronym{tp}{TP}{True Positive}
\newacronym{fp}{FP}{False Positive}
\newacronym{fn}{FN}{False Negative}

\newcommand{\figref}[2]{Fig.~\hyperref[#1]{\ref*{#1}#2}}
\newcommand{\edfigref}[2]{Extended Data Fig.~\hyperref[#1]{\ref*{#1}#2}}
\newcommand{\edfigrefrg}[3]{Extended Data Fig.~\hyperref[#1]{\ref*{#1}#2-\ref*{#1}#3}}

\graphicspath{{../figures}}
%% my package end


%%%%%=============================================================================%%%%
%%%%  Remarks: This template is provided to aid authors with the preparation
%%%%  of original research articles intended for submission to journals published
%%%%  by Springer Nature. The guidance has been prepared in partnership with
%%%%  production teams to conform to Springer Nature technical requirements.
%%%%  Editorial and presentation requirements differ among journal portfolios and
%%%%  research disciplines. You may find sections in this template are irrelevant
%%%%  to your work and are empowered to omit any such section if allowed by the
%%%%  journal you intend to submit to. The submission guidelines and policies
%%%%  of the journal take precedence. A detailed User Manual is available in the
%%%%  template package for technical guidance.
%%%%%=============================================================================%%%%

%% as per the requirement new theorem styles can be included as shown below
\theoremstyle{thmstyleone}%
\newtheorem{theorem}{Theorem}%  meant for continuous numbers
%%\newtheorem{theorem}{Theorem}[section]% meant for sectionwise numbers
%% optional argument [theorem] produces theorem numbering sequence instead of independent numbers for Proposition
\newtheorem{proposition}[theorem]{Proposition}%
%%\newtheorem{proposition}{Proposition}% to get separate numbers for theorem and proposition etc.

\theoremstyle{thmstyletwo}%
\newtheorem{example}{Example}%
\newtheorem{remark}{Remark}%

\theoremstyle{thmstylethree}%
\newtheorem{definition}{Definition}%

\raggedbottom
\unnumbered% uncomment this for unnumbered level heads

\begin{document}

\title[Article Title]{A Genomic Language Model for Chimera Artifact Detection in Nanopore Direct RNA Sequencing}

%%=============================================================%%
%% GivenName	-> \fnm{Joergen W.}
%% Particle	-> \spfx{van der} -> surname prefix
%% FamilyName	-> \sur{Ploeg}
%% Suffix	-> \sfx{IV}
%% \author*[1,2]{\fnm{Joergen W.} \spfx{van der} \sur{Ploeg}
%%  \sfx{IV}}\email{iauthor@gmail.com}
%%=============================================================%%

\author[1]{\fnm{Yangyang} \sur{Li}}\email{yangyang.li@northwestern.edu}
\equalcont{These authors contributed equally to this work.}

% \author*[1,2]{\fnm{First} \sur{Author}}\email{iauthor@gmail.com}
\author[1]{\fnm{Tingyou} \sur{Wang}}\email{tywang@northwestern.edu}
\equalcont{These authors contributed equally to this work.}

\author[1]{\fnm{Qingxiang} \sur{Guo}}\email{qingxiang.guo@northwestern.edu}

\author[1]{\fnm{Yanan} \sur{Ren}}\email{ynren1020@gmail.com}
\author[1]{\fnm{Xiaotong} \sur{Lu}}\email{xiaotong.lu@northwestern.edu}

\author[1,2]{\fnm{Cao} \sur{Qi}}\email{qi.cao@northwestern.edu}
\author*[1,2]{\fnm{Rendong} \sur{Yang}}\email{rendong.yang@northwestern.edu}

\affil[1]{\orgdiv{Department of Urology}, \orgname{Northwestern University Feinberg School of Medicine}, \orgaddress{\street{303 E Superior St}, \city{Chicago}, \postcode{60611}, \state{IL}, \country{USA}}}
\affil[2]{\orgdiv{Robert H. Lurie Comprehensive Cancer Center}, \orgname{Northwestern University Feinberg School of Medicine}, \orgaddress{\street{675 N St Clair St}, \city{Chicago}, \postcode{60611}, \state{IL}, \country{USA}}}

% \author[1,2]{\fnm{Third} \sur{Author}}\email{iiiauthor@gmail.com}
% \equalcont{These authors contributed equally to this work.}

% \affil*[1]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \city{City}, \postcode{100190}, \state{State}, \country{Country}}}
% \affil[2]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \city{City}, \postcode{10587}, \state{State}, \country{Country}}}
% \affil[3]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \city{City}, \postcode{610101}, \state{State}, \country{Country}}}

%%==================================%%
%% Sample for unstructured abstract %%
%%==================================%%

\abstract{
	Nanopore \gls{drs} has revolutionized transcriptomics but is challenged by artificial chimeric reads that compromise data integrity.
	We present DeepChopper, a novel large language model tailored for biological sequences, which accurately identifies and removes artificial sequences in nanopore \gls{drs} data without relying on alignment information.
	DeepChopper includes long-range dependency modeling with quality-aware processing and achieves both broad context understanding and single-nucleotide resolution.
        Across multiple cell lines and sequencing platforms, DeepChopper reduced chimeric reads by 62-84\% and improved supporting rates from 8-19\% to 43-55\% compared to existing methods.
	In particular, in gene fusion detection, DeepChopper reduced false positives by 89\% while increasing the proportion of supported fusions from 2\% to 17\%.
	By improving data quality, DeepChopper significantly improves the reliability of downstream analyses, particularly in cancer genomics and transcriptomics.
	This work demonstrates the powerful potential of large language models in analyzing complex biological data, paving the way for advancements in genomics and biotechnology.}
\maketitle

\section{Main}

% Outline
% 1. Sequencing history
% 2. Genomics LLM history
% 3. Challenges of LLM / efficiency of transformer and improvements
% 3.1 long range
% 3.2 single nucleotide resolution
% 4. Cons of drs
% 5. Our method
% 6. Results
% 7. Conclusion

Long-read RNA sequencing technologies are revolutionizing transcriptomic research by providing unparalleled resolution for detecting complex splicing and gene fusion events often missed by conventional short-read RNA-seq methods.
Among these technologies, \gls{ont} \gls{drs} stands out by sequencing full-length RNA molecules directly, preserving native RNA modifications and allowing a more accurate and comprehensive analysis of RNA biology.
This approach bypasses the inherent limitations of cDNA-based long-read sequencing methods, such as artifacts arising from reverse transcription, template switching, and \gls{pcr} amplification~\cite{garalde2018highly, jain2022advances}.

Despite these advantages, a critical question remains: Does \gls{ont} \gls{drs} introduce technical artifacts?
A previous study has suggested that \gls{drs} might generate chimera artifacts, leading to multi-mapped reads~\cite{smith2020molecular}.
These artifacts may result from ligation during library preparation or chimeric reads produced by software in silico, potentially confounding downstream analyses such as transcriptome assembly, quantification, and detection of alternative splicing and gene fusion events.
Detecting these chimera artifacts is challenging because long-read aligners often produce chimeric alignments from such artifacts that are indistinguishable from those derived from true gene fusion events.
Importantly, chimeric read artifacts frequently contain internal adapter sequences~\cite{smith2020molecular}, which could theoretically serve as a distinguishing feature to differentiate them from true gene fusion-derived reads. 
However, \gls{ont} \gls{drs} basecallers, trained in RNA, struggle to properly call these DNA-based adapter sequences under an RNA model~\cite{liu2024sequencing}.
As a result, current adapter detection tools cannot exploit this feature to eliminate chimeric read artifacts, leaving the issue unresolved.

To address these challenges, we developed DeepChopper, a \gls{glm} specifically designed for long-read sequence analysis.
Inspired by recent advances in \gls{llm} that can decipher complex patterns in genetic sequences~\cite{benegas2024genomic}, DeepChopper excels at processing long genomic contexts with single-nucleotide resolution, a crucial capability for identifying ONT adapter sequences directly from base-called long reads and thus facilitating the detection and removal of chimeric read artifacts in \gls{drs} data.
Specifically, DeepChopper processes individual FASTQ records to identify and handle adapter sequences: When internal adapters are detected, the model splits the read into two or more records, while 3' end adapters are trimmed to produce a clean sequence (\figref{fig:f1}{a}).
This method ensures the preservation of genuine biological sequences while effectively eliminating chimera artifacts. 

To accomplish this, DeepChopper combines the \gls{lcglm} HyenaDNA~\cite{nguyen2024hyenadna}, optimized for efficient long-range dependency modeling, with a residual block, and \gls{mlp} for fine-grained sequence analysis.
This hybrid architecture enables both broad context understanding and precise single-nucleotide resolution~\cite{poli2023hyena, he2016deep} (\figref{fig:f1}{a}).
Additionally, DeepChopper incorporates base quality score information to improve prediction accuracy by accounting for sequencing confidence, and employs an adaptive tokenization strategy that preserves single-nucleotide resolution while capturing higher-order sequence patterns (See \hyperref[sec:methods]{Methods} for details).
DeepChopper offers several advantages over existing genomic language models.
Unlike DNABERT2~\cite{zhou2023dnabert2}, which is limited to a context length of 512 base pairs, DeepChopper processes up to 32 kilobase pairs, a sufficiently long context to capture most mRNAs.
Additionally, unlike Nucleotide Transformer~\cite{dalla2023nucleotide}, which lacks single-base resolution, DeepChopper can accurately identify non-human reference sequences, such as \gls{ont} adapters, with base-pair precision. 
Moreover, its efficient architecture—using only 7 million parameters—makes DeepChopper computationally practical for large-scale \gls{drs} analysis, in contrast to resource-intensive models like Evo~\cite{nguyen2024sequence}, which rely on billions of parameters and may not be feasible in environments with limited computational resources.
This combination of long context processing, high-resolution accuracy, and computational efficiency makes DeepChopper particularly well-suited for analyzing massive long-read sequences.

% Result
\begin{figure}[!h]
	\includegraphics[height=0.78\columnwidth]{finals/figure1}
	\caption{{\bf  Detection and analysis of chimeric reads in nanopore \gls{drs} data using DeepChopper.} (a) Overview of the architecture of the DeepChopper model. Created with BioRender.com (b) Length distribution of predicted artifacts in the VCaP 002 dataset. (c) Relative position of the prediction artifacts in the VCaP 002 dataset. Top: Schematic illustrating the calculation of the relative position (rp). Grey rectangle represents a read from 5$'$ to 3$'$. Red arrow indicates the chopped site, marking the starting position of the chopped intervals. Relative position is calculated as the ratio of the length before the chopped site to the total read length. Bottom: Distribution of the relative position of the prediction artifacts in the dataset. (d) Count of remaining intervals per read in the VCaP 002 dataset. (e) Chimeric read detection of DeepChopper compared to Dorado (with and without trimming) in the VCaP 002 dataset. The grey bars represent chimeric reads unsupported by \gls{ont} \gls{pcr} cDNA, while the red bars indicate chimeric reads supported by \gls{ont} \gls{pcr} cDNA.  (f) The base quality score of the artificial sequences identified by DeepChopper in the VCaP 002 dataset. Background colors indicate quality levels: green (high), yellow (medium), and red (low). (g) \gls{blat} identity analysis of the artificial sequences detected by DeepChopper in the VCaP 002 dataset. (h) Chimeric read detection of DeepChopper compared to Dorado with trimming across multiple cell lines, including A549, HCT116, HepG2, K562, and MCF7. In each sample, the grey bars represent chimeric reads unsupported by \gls{ont} \gls{pcr} cDNA, while the red bars indicate chimeric reads supported by \gls{ont} \gls{pcr} cDNA. (i)  Supporting read analysis for WTC11, calculated the support rate using five different datasets: \gls{ont} \gls{pcr} cDNA, \gls{ont} CapTrap-seq, \gls{ont} \gls{r2c2}, \gls{pb} cDNA, and \gls{pb} CapTrap-seq. The blue bars represent unsupported chimeric reads, while the green bars indicate supported chimeric reads}\label{fig:f1}
\end{figure}

% Analysis of Predicted Artifact Lengths Fig 1b
To train DeepChopper for identifying adapter sequences within dRNA-seq long reads, we utilized data from six human cell lines: HEF293T, A549, HCT116, HepG2, K562 and MCF-7 provided by the Singapore Nanopore Expression Project (SG-NEx)~\cite{chen2021systematic}.
We curated a training set of 540,000 long reads initially deemed free of adapters and inserted putative adapter sequences, derived from the raw dRNA-seq data, into these reads to create instances of internal and 3' end adapters (See \hyperref[sec:methods]{Methods} for details).
An independent test set comprising 60,000 long reads was held out for performance evaluation.
DeepChopper achieved recall, precision, and F1 scores above 0.99 (\edfigref{fig:sf1}{a}), demonstrating its high accuracy in detecting adapter sequences.
To further assess DeepChopper’s ability to detect chimera artifacts in real data, we generated an additional \gls{drs} dataset using the prostate cancer VCaP cell line, which was not included in the training data.
This independent dataset provides a robust framework for evaluating chimera artifacts in genuine \gls{drs} samples, ensuring that DeepChopper's performance generalizes beyond the training data. 

We conducted \gls{drs} of VCaP cells using \gls{ont}'s SQK-RNA002 chemistry, consistent with that used in the SG-NEx project.
Using a MinION sequencer with three R9.4 flow cells, we generated a total of 9,177,639 long reads in FASTQ format, with base-calling performed by \gls{ont}'s Dorado software~\cite{dorado2023}.
DeepChopper was then applied for adapter trimming and correction of chimeric read artifacts.
In particular, DeepChopper increased the yield of long-read sequences by 3\%, resulting in a final output of 9,357,913 long reads.
DeepChopper identified 8,218,172 adapter sequences within 7,990,102 long reads, accounting for 87\% of the raw reads. 
Most of these adapter sequences were approximately 70 bp in length (\figref{fig:f1}{b}), aligning with the expected length of the RMX adapter used in \gls{ont} SQK-RNA002 \gls{drs} kit, as specified in the kit's technical documentation~\cite{nano2017tech}.

We then analyzed the positions of the adapter sequences within the long reads identified by DeepChopper.
Among these, 212,478 reads contained internal adapters, while 7,777,624 reads had adapters at their 3' ends (\figref{fig:f1}{c}).
This reveals that chimeric read artifacts, evidenced by internal adapters, are abundant in VCaP \gls{drs} data and may reflect a common problem inherent to \gls{drs} long reads.
Further examination of adapter-trimmed reads showed that chimeric artifacts could arise from the joining of multiple long reads, with the most prevalent type involving two reads connected by an internal adapter to form a single chimeric read (\figref{fig:f1}{d}).

To validate the chimeric read artifacts identified by DeepChopper in the VCaP \gls{drs} data, we analyzed the chimeric alignments generated by Minimap2\cite{li2018minimap2}, as chimeric reads typically produce such alignments.
To distinguish genuine chimeric reads from artifacts, we generated Nanopore direct cDNA sequencing data for VCaP cells and considered chimeric reads with alignments fully supported by cDNA sequencing as bona fide (See \hyperref[sec:methods]{Methods} for details).
Additionally, we assessed whether the adapter trimming function of \gls{ont}'s Dorado basecaller could mitigate chimeric artifacts.
Although Dorado with the trimming option did not reduce the chimeric alignments, DeepChopper reduced them by 91\% and increased the proportion of cDNA-supported chimeric alignments from 5\% to 47\% (\figref{fig:f1}{e}).
To further confirm the nature of the identified chimeric artifacts, we evaluated the base quality and aligned DeepChopper-identified internal adapter sequences against the human reference genome using \gls{blat}~\cite{kent2002blat}, which offers high accuracy local alignment.
We found that adapter sequences from identified chimeric read artifacts exhibited low base quality scores (\figref{fig:f1}{f}) and alignment identity (\figref{fig:f1}{g}), confirming their artifactual nature and non-human origin.
These results demonstrate that DeepChopper effectively addresses \gls{drs} chimera artifacts, a challenge that current tools cannot resolve.

To comprehensively assess DeepChopper’s performance in reducing chimera artifacts in \gls{drs}, we examined cDNA-supported chimeric alignments in the SG-NEx samples. 
DeepChopper significantly reduced chimeric alignments by 62\% to 84\%, while preserving cDNA-supported chimeric alignments without noticeable reduction (\figref{fig:f1}{h}), indicating the prevalence of chimeric read artifacts in \gls{drs}. 

To further validate the accuracy of DeepChopper, we applied it to long-read data from the WTC11 human cell line generated by the LRGASP project~\cite{pardo2024systematic}, which included ONT dRNA-seq and various cDNA-based long-read sequencing protocols (PCR-cDNA, CapTrap, \gls{r2c2}) from the ONT and PacBio platforms. DeepChopper consistently eliminated only the chimeric alignments in the \gls{drs} that lacked support from any cDNA-based sequencing method (\figref{fig:f1}{i}), demonstrating its precise ability to distinguish true chimeric reads from artifacts.

Recently, \gls{ont} released a new SQK-RNA004 chemistry for \gls{drs}, but it remains unclear whether chimera artifacts persist with this update. 
To investigate, we generated \gls{drs} data using the latest SQK-RNA004 kit for the VCaP cell line and performed zero-shot predictions of chimera artifacts using DeepChopper, despite the adapter sequence pattern of RNA004 not being included in its training.
DeepChopper reduced chimeric alignments by 21\% compared to raw alignments from Dorado’s base-called and adapter-trimmed long reads, increasing the proportion of cDNA-validated chimeric alignments from 25\% to 30\% (\edfigref{fig:sf1}{b}).
Furthermore, the internal adapter sequences identified by DeepChopper within these RNA004 chimeric read artifacts exhibited low base quality scores (averaging around 10) and low \gls{blat} identity (clustering around 0.3) (\edfigref{fig:sf1}{c}).
These findings strongly suggest that the chimera artifact issue continues to be present in the latest RNA004 \gls{drs} kit.

\begin{figure}[H]
	\includegraphics[height=1\columnwidth]{finals/figure2}
	\caption{{\bf Characterization of \gls{fp} chimeric reads Detected by DeepChopper in the VCaP 002 dataset.} (a) Chromosomal distribution and connections in \gls{fp} chimeric reads of VCaP 002. Top: Bar plot showing the number of connections per base pair for each chromosome. Higher bars indicate more frequent connections. Bottom: Heatmap illustrating the count of connections between chromosomes. Intensity of color corresponds to the number of connections. (b) Comparison of isoforms identified in DeepChopper and Dorado processed data. After DeepChopper treatment, an increased number of isoforms were observed across all categories except for NNC. (c) Gene fusion detection in VCaP 002 dataset. Comparison of DeepChopper and Dorado (with trimming) performance. The grey bars represent gene fusions unsupported by RNA-seq data, while the red bars indicate gene fusions supported by RNA-seq data. (d) \gls{go} enrichment of for the chimeric artifacts in VCaP 002. (e) Fusion of RPS29 and COX8A identified from artificial chimeric reads from the VCaP 002 dataset. The polyA sequence within the RNA read is highlighted in red, and regions trimmed by DeepChopper are indicated with grey dashed lines. Accompanying the read sequence are base quality scores and normalized ONT ion current signals. Red circles mark a brief ``open pore'' state that Dorado missed, preventing the accurate segmentation of the signal into two independent reads.}\label{fig:f2}
\end{figure}

To assess the impact of chimera artifacts on downstream analyses, we examined the patterns of improper connections generated by chimera artifacts based on their genomic locations across chromosomes in the VCaP \gls{drs} data. We observed varying frequencies of artificial connections across chromosomes, with the mitochondrial chromosome (Chr M) exhibiting a particularly high rate of connections per base pair, suggesting it may be a hotspot for chimera artifact formation (\figref{fig:f2}{a}). Additionally, a similar pattern of improper connections between chromosomes was observed in the VCaP RNA004 \gls{drs} data (\edfigref{fig:sf2}{a}), suggesting that the source of chimera artifacts may be intrinsic to the \gls{drs} technology or related to certain genomic features, rather than being entirely mitigated by changes in sequencing chemistry. In addition, we found that genes affected by chimera artifacts exhibited significantly higher expression levels compared to all expressed genes (\(\textrm{p-value} < 2.2 \times 10^{-16}\)) (\edfigref{fig:sf2}{b}), while following a similar size distribution pattern (\edfigref{fig:sf2}{c}).

Given DeepChopper's ability to increase the yield of sequencing reads by correcting chimera artifacts, we reasoned that it could enhance transcript detection and quantification. Using IsoQuant [citation] for long-read transcript annotation, we found that DeepChopper nearly doubled the number of identified transcripts compared to Dorado adapter-trimmed long-reads (\figref{fig:f2}{b}). This improvement was particularly evident in the detection of full-length transcripts, as shown by the increase in the full splice match (FSM) category (\figref{fig:f2}{b}). Additionally, DeepChopper enhanced the discovery of alternatively spliced transcripts, including those classified as ISM (incomplete splice match), NIC (novel in catalog), and NNC (novel not in catalog), further demonstrating its effectiveness in improving transcript diversity and accuracy in quantification.

% Gene Ontology (GO) Enrichment Analysis of FP Group Genes (Fig 2b)
To further explore the functional functions of genes in the \gls{fp} group, we conducted a \gls{go} enrichment analysis in six cell lines: VCaP, A549, K562, HepG2, MCF7, and HCT116 (\edfigrefrg{fig:sf2}{d}{i}).
We illustrate the overlapping \gls{go} pathways enriched among the \gls{fp} genes across these cell lines (\figref{fig:f2}{c}).
The results show a substantial overlap in \gls{go} terms related to core biological processes, particularly those associated with ribosomal small subunit biogenesis, cytoplasmic translation, translational elongation, and rRNA processing (\figref{fig:f2}{c}).
This overlap indicates that genes in the \gls{fp} group are frequently involved in pathways critical to protein synthesis and ribosome function.
Such pathways are typically highly expressed and conserved in different cell types, which might explain their recurrent identification as false positives.
The significant intersection sizes for these \gls{go} terms in the six cell lines suggest a common pattern in the \gls{fp} gene set, strengthening the hypothesis that these genes might be artifacts associated with sequencing or data processing errors rather than true biological signals.
This finding highlights the need for careful evaluation in high-throughput sequencing studies to avoid misleading conclusions about gene function.

% (Fig 2c)
DeepChopper processed data shows a higher number of isoforms, particularly in the FSM category, indicating its superior ability to accurately trim full-length transcripts from artificial chimeras (\figref{fig:f2}{c}).


% Gene fusion detection in VCaP 002 dataset.  (Fig 2d)
We compared gene fusion detection in the VCaP 002 dataset processed by DeepChopper and Dorado with trimming (\figref{fig:f2}{d}).
DeepChopper significantly outperformed Dorado, reducing total gene fusion events by 89\% and increasing the proportion of supported fusions (validated by RNA-seq data) from 3\% (Dorado) to 33\% (DeepChopper)  (\figref{fig:f2}{d}).
This marked improvement in precision demonstrates the ability of DeepChopper to distinguish genuine gene fusions from artifacts.
This precision is crucial for applications in cancer genomics and transcriptomics, where correct identification of gene fusions is critical.

% The circular plot shows an overview of chromosal chonnection of fusion events detected in the FP chimeric reads in VCaP 002 dataset between locations in chromosomes. (sumplementary fig 3)
\edfigref{fig:sf3}{} visualizes the chromosomal connections of the fusion events in \gls{fp} chimeric reads from the VCaP 002 dataset.
The circular plot reveals fusion events across all chromosomes, including sex chromosomes and mitochondrial DNA.
Red and blue links indicate intrachromosomal and interchromosomal fusions, respectively.
In particular, ribosomal protein genes (RPL and RPS families) are frequently involved in these fusion events, suggesting a potential bias towards highly expressed genes in artificial chimeric read formation.
Recurrent fusion partners, such as RPL11, RPS8, and RPL31, emerge as potential hotspots. This diverse pattern of chromosomal connections underscores the complexity of artificial chimeras in nanopore direct RNA sequencing data and highlights the necessity for advanced detection methods like DeepChopper to accurately identify and filter these events.

% Example of a misleading gene fusion caused by an artificial chimeric read (Fig 2e)
False positive fusions, such as RPS29-COX8A, were identified through the clear visual confirmation of dual adapters in the raw signal. The presence of multiple adapters within a single read may arise from free-floating adapters in solution and insufficient time between the passage of the first adapter and the next true read. 
Furthermore, the lack of a distinct open pore signal can cause Dorado to miss proper segmentation, resulting in a single POD5 file that captures both events, further complicating the analysis (\figref{fig:f2}{e}).

In conclusion, DeepChopper demonstrates the powerful potential of \gls{llm} in the analysis of long biological sequences, particularly in nanopore \gls{drs}.
By accurately identifying and removing artificial sequences, including internal adapters, DeepChopper significantly enhances the quality and reliability of sequencing data.
This approach not only improves the current nanopore \gls{drs} analysis, but also showcases the broader applicability of language models to complex biological data.
Our work paves the way for further advancements in genomics and transcriptomics, suggesting exciting possibilities for optimizing these models for specific biological tasks.
As large language models continue to evolve, their role in biotechnology and molecular biology is likely to expand, promising new frontiers in the analysis and interpretation of genetic information.

\section{Methods}\label{sec:methods}

\subsection{Cell culture}

VCaP cell line was obtained from the American Type Culture Collection (ATCC) and cultured under sterile conditions to maintain optimal growth and viability.
The cells were grown in Dulbecco's Modified Eagle Medium (DMEM, high glucose; Gibco, Cat\# 11-965-092) supplemented with 10\% fetal bovine serum (FBS Opti-Gold, Performance Enhanced, US Origin; Gendepot, Cat\# F0900-050) to provide essential growth factors.
In addition, the culture medium was enriched with \SI{5}{\ml} of 100 mM Sodium Pyruvate (Gendepot, Cat\# CA017-010) to support cellular metabolism and \SI{5}{\ml} of Antibiotics-Antimycotics (\( 100\times \)) (Gendepot, Cat\# CA002-010) to prevent microbial contamination.
Cells were cultured in \SI{100}{\mm} cell culture treated dishes (Thermo Fisher Scientific, Cat\# 12-556-002) and incubated at \SI{37}{\degreeCelsius} in a humidified atmosphere containing 5\% CO2, with media changes performed every 72 hours to ensure nutrient availability and waste removal.
Cell confluency was regularly monitored and subculturing was performed before reaching 80\% confluency to maintain healthy growth conditions and prevent over-confluence stress.

\subsection{RNA extraction and quantification}

Total RNA was extracted using the RNeasy Mini Kit (Qiagen, Cat\# 74104) according to the protocol of the manufacturer.
The quality and concentration of RNA were assessed using an Agilent 2100 Bioanalyzer.
Poly(A)+ RNA was then enriched from total RNA using the Dynabeads\textsuperscript{tm} mRNA Purification Kit (Invitrogen, Cat\# 65001), which utilizes oligo (dT) beads for selective mRNA binding.
The mRNA was quantified using a Qubit 4 fluorometer and a Qubit RNA HS Assay Kit (Thermo Fisher Scientific, Cat\# Q32852).
The mRNA preparations were either immediately used to prepare a sequencing library or frozen and stored at \SI{-80}{\degreeCelsius} until further use.

\subsection{Nanopore sequencing}

We performed Direct-RNA nanopore sequencing of the enriched mRNA using two different sets: the RNA002 kits with R9.4.1 flow cells and the RNA004 kits with R10.4.1 flow cells.
The decision to incorporate the RNA004 kit, a newly released option, was driven by our intention to test its capabilities in conjunction with our DeepChopper tool to optimize data quality and sequencing efficiency.
For the RNA002 library, \SI{1}{\micro\gram} of poly(A)+ RNA was used as input for library preparation using the Direct RNA Sequencing Kit (SQK-RNA002, \gls{ont}) following the manufacturer's instructions.
Nanopore \gls{drs} employs a reverse transcriptase adapter (RTA) that typically binds to the poly(A) tails of messenger RNA (mRNA); subsequently, a sequencing adapter is ligated to the RTA, which guides the mRNA through the nanopore for sequencing.
The prepared library was loaded onto a MinION R9.4 flow cell (FLO-MIN106) and sequenced for 48 hours using the Oxford Nanopore MinION device.
For the RNA004 library, 300 ng of poly(A)+ RNA was used as input for library preparation using the Direct RNA Sequencing Kit (SQK-RNA004, \gls{ont}) according to the protocol of the manufacturer.
The library was then loaded onto a PromethION RNA Flow Cell (FLO-PRO004RA) and sequenced on the Oxford Nanopore PromethION device for 72 hours.

For Direct cDNA sequencing, we utilized the Direct cDNA Sequencing Kit (SQK-DCS109, \gls{ont}) following the manufacturer's protocol.
Briefly, \SI{5}{\micro\gram}  of total RNA was used as input for first-strand cDNA synthesis using Maxima H Minus Reverse Transcriptase (Thermo Fisher Scientific) with the SSP and VN primers provided in the kit.
To eliminate potential RNA contamination, we treated the sample with RNase Cocktail Enzyme Mix (Thermo Fisher Scientific).
Second-strand cDNA synthesis was carried out using LongAmp Taq Master Mix (New England Biolabs).
The resulting double-stranded cDNA underwent end-repair and dA-tailing using the NEBNext Ultra End Repair/dA-Tailing Module (New England Biolabs).
Subsequently, sequencing adapters were ligated to the prepared cDNA using Blunt/TA Ligase Master Mix (New England Biolabs).
Between each enzymatic step, the cDNA and libraries were purified using AMPure XP beads (Agencourt, Beckman Coulter).
We quantified the libraries using a Qubit Fluorometer 3.0 (Life Technologies) to ensure adequate concentration and quality.
The final library was loaded onto a MinION R9.4 flow cell and sequenced on the Oxford Nanopore MinION device for 72 hours.

\subsection{Training data preparation}\label{ssec:data}

We obtained \gls{ont} directRNA FAST5 data from the SG-Nex project, which includes six cell lines (HEK293T, A549, K562, HepG2, MCF7, and HCT116)~\cite{chen2021systematic}, and subsequently converted the data to POD5 format using the POD5 tool (\url{https://pod5-file-format.readthedocs.io}).
FASTQ files were generated using Dorado (v0.5.2)~\cite{dorado2023} with adapter trimming disabled (--no-trim) and  model ``rna002\_70bps\_hac@v3''.
The reads were aligned to the human reference genome (GRCh38) using minimap2 (v2.24)~\cite{li2018minimap2} with optimized \gls{ont} directRNA parameters (-ax splice -uf -k14).
For adapter extraction, we selected primary alignments lacking supplementary alignments, treating \(3^{\prime}\)  soft-clipped regions as adapter sequences and aligned portions as nonadapter sequences.

To create artificial chimeric reads, we randomly combined two non-adapter sequences with one adapter sequence.
The resulting SAM files were converted to BAM format, indexed, and sorted using SAMtools (v1.19.2)~\cite{li2009sequence}.
Our dataset comprises both positive (terminal and internal adapters in a 1:1 ratio) and negative examples (no artificial sequences) in a 9:1 ratio.
We generated 600,000 data points, divided into training (480,000), validation (60,000) and test (60,000) sets with a ratio of 8:1:1.

\subsection{Language model architecture}\label{ssec:lm}

DeepChopper approaches artificial sequence detection as a token classification task, predicting whether each token belongs to an artificial sequence.
The model architecture is specifically designed for this purpose, allowing for accurate identification of artificial sequence regions within biological data.

At its core, DeepChopper adapts the HyenaDNA framework, a genomics model architecture, as its primary feature extractor.
This choice enhances the ability of the model to process and interpret complex biological sequences effectively.
Following feature extraction, we integrated a quality block to incorporate sequence quality information into the prediction process.
This block, comprising dense layers with residual connections, combines quality scores with extracted features, enriching the model's understanding of the input data.

The integrated features are then passed to a classification head, which outputs the probability that each token is part of an artificial sequence.
This architecture allows DeepChopper to leverage contextual information within biological sequences, ensuring high accuracy in artificial sequence prediction.
By classifying tokens into positive (part of an artificial sequence) or negative (not part of an artificial sequence) categories, DeepChopper provides a nuanced analysis of the input data.


\subsection{Model training}\label{ssec:training}

We used a tokenization strategy that converted the biological sequences into single-nucleotide tokens.
The tokens contain \emph{A}, \emph{C}, \emph{G}, \emph{T} and \emph{N} as the basic units, allowing the model to capture the fine-grained details of the sequences.
DeepChopper processes sequences up to 32,770 nucleotides in length, with longer sequences excluded from analysis.
For efficient batch processing, shorter sequences are padded to match this maximum length.
The model was trained using supervised learning, with sequences annotated for artificial sequences.
Training was carried out in a \gls{hpc} cluster using two A100 \glspl{gpu}.
The batch size was 64, the validation was performed every 20,000 steps, and we selected the model with the highest validation F1 score for the base prediction task.
The Adam optimizer was used for training with default parameters \( \beta_{1} = 0.9 \) and \( \beta_{2} = 0.999 \)~\cite{kingma2014adam}.
We use a learning rate scheduler to reduce the learning rate when the validation loss stops improving, and the initial learning rate is \( 2 \times 10^{-5} \).

Cross-entropy loss was used to update the model parameters.

\[
	\ell(x, y) = L = \{l_1,\cdots,l_N\}^\top, \quad
	l_n = - w_{y_n} \log \frac{\exp(x_{n,y_n})}{\sum_{c=1}^C \exp(x_{n,c})}
	\cdot 1\{y_n \not= \textrm{ignore\_index}\}
\]

where \( x \) is the input, \( y \) is the target, \( w \) is the weight,
\( C \) is the number of classes, and \( N \) spans the dimension of the mini-batch and
\( d_1, \cdots, d_k \) for the K-dimensional case.

\[
	\ell(x, y) =   \sum_{n=1}^N \frac{1}{\sum_{n=1}^N w_{y_n} \cdot 1\{y_n \not= \textrm{ignore\_index}\}} l_n
	.\]

Training was carried out for \num{60} epochs with early stopping based on validation performance to prevent overfitting.
Evaluation metrics included accuracy, precision, recall, and F1 score with the following equations:
\begin{align*}
	\textrm{Precision} & = \frac{\textrm{TP}}{\textrm{TP}+\textrm{FP}}                                                     \\
	\textrm{Recall}    & = \frac{\textrm{TP}}{\textrm{TP}+\textrm{FN}}                                                     \\
	\textrm{F1}        & = 2 \times \frac{\textrm{Precision} \times \textrm{Recall}}{\textrm{Precision} + \textrm{Recall}}
\end{align*}

The final model was selected on the basis of the best performance in the validation set.
To find the optimal model parameters, we used Hydra~\cite{Yadan2019Hydra}.
The total loss for one window is the sum of the losses for each supported position.
For per position \( i \), where \( i  \in {1, 2, \cdots, S} \),  the total loss includes both base prediction loss.
The binary cross-entropy losses were used for these tasks, with the total loss incorporating predicted base probabilities, true base values, and the probability that a position is informative.
The complete list of hyperparameters can be found in the Supplementary Table~\ref{tab:hyperparameter}.

\subsection{Polishing predictions}

To enhance the accuracy and smoothness of the predictions, we implemented a sliding window approach.
This technique helps to extend prediction regions and reduces choppy noise, aligning with the typical distribution of artificial sequences in nanopore \gls{drs} data.
The sliding window technique is particularly effective in maintaining continuity in regions where artificial sequences are predicted, ensuring that isolated predictions do not create fragmented and unrealistic results.

We used a sliding window of size 21 nucleotides.
This window size was chosen based on empirical testing to balance the need for sufficient smoothing while retaining sensitivity to shorter artificial sequences\qx{citation like Valeri, Jacqueline A., Katherine M. Collins, Pradeep Ramesh, Miguel A. Alcantar, Bianca A. Lepe, Timothy K. Lu, and Diogo M. Camacho. "Sequence-to-function deep learning frameworks for engineered riboregulators." Nature communications 11, no. 1 (2020): 5058.} .

Within each sliding window, we applied a voting method to polish the predictions.
The final classification of each nucleotide within the window is determined by the majority vote of all predictions within that window.
The voting result for each nucleotide \( x_i \) is given by:
\[
	y_i = \begin{cases}
		1 & \text{if } \sum_{j=i-k}^{i+k} p_j > \frac{W}{2} \\
		0 & \text{otherwise}
	\end{cases}
\]
where \( y_i \) is the final prediction for nucleotide \( x_i \). \( p_j \) is the predicted probability of being an artificial sequence for the nucleotide \( x_j \). \( W \) is the size of the sliding window, \( k \) is half the window size, i.e., \( k = \frac{W-1}{2} \).

Following prediction refinement using a sliding window approach, we applied additional filters to enhance output quality.
Predicted artificial sequences shorter than 13 nucleotides were removed to minimize false positives due to noise.
Reads containing more than four predicted artificial sequences were excluded from the processing.
The filtered sequences and their associated quality scores were then outputted as FASTQ records.

This post-processing step ensures that the final output is accurate and biologically relevant.
The use of sliding windows and the voting method significantly improved the quality of the predictions by smoothing the output and reducing false positives.
By combining these techniques, we were able to produce more reliable predictions, which are crucial for downstream applications.


\subsection{Evaluating artificial sequences detection}

We evaluated DeepChopper performance using precision, recall and the F1 score in the test data.
These metrics provided a comprehensive view of the model's ability to identify artificial sequences.
To evaluate DeepChopper's effectiveness across diverse datasets, we analyzed its performance on multiple cell lines (VCaP 002, A549, HCT116, K562, MCF7) and compared it to baseline methods, including Dorado with and without trimming.
We quantified the reduction in chimeric reads and the improvement in supporting rates across these datasets.

We further validated DeepChopper's robustness using cross-platform analysis.
Using WTC11 cell lines, we calculated supporting read rates using data from five different sequencing platforms: \gls{ont} \gls{pcr} cDNA, \gls{ont} CapTrap-seq, \gls{ont} \gls{r2c2}, \gls{pb} cDNA, and \gls{pb} CapTrap-seq.
This cross-platform validation assessed the ability of DeepChopper to maintain high support rates under various sequencing conditions.
To evaluate the quality of the detected artificial sequences, we analyzed their base quality score and \gls{blat} identity.
We compared these metrics for artificial sequences detected in Dorado-unique \gls{fp} chimeric reads with trimming versus DeepChopper.
The model's generalizability was further validated by testing on independent datasets, including the VCaP 004 dataset sequenced using the newer RNA004 protocol, demonstrating DeepChopper's adaptability to evolving sequencing technologies.

In addition, we used the VCaP 002 dataset, focusing on identifying chimeric reads caused by internal adapters.
We cross-referenced the identified reads with data from multiple sequencing platforms: \gls{ont} \gls{pcr} cDNA, \gls{ont} CapTrap-seq, \gls{ont} \gls{r2c2}, \gls{pb} cDNA, and \gls{pb} CapTrap-seq.
Supporting rates, defined as the ratio of chimeric reads corroborated by other platforms to the total detected chimeric reads, were calculated to assess the detection consistency across platforms.
We extended our evaluation to multiple cell lines (A549, HCT116, HepG2, K562, MCF7) and WTC11 cell lines to demonstrate the broad applicability of DeepChopper.
This cross-sample and cross-platform validation approach allowed us to assess the model's performance across diverse sequencing contexts.

\textit{BLAT Identity Calculation.} To assess the accuracy of the chimeric read artificial sequences, we calculated the \gls{blat} identity for each sequence using both \gls{blat} (v0.37.2)~\cite{kent2002blat} and PxBLAT (v1.2.1)~\cite{li2024pxblat}.
The BLAT identity is defined as the ratio of match length to sequence length, providing a measure of how closely a sequence aligns to a reference genome.
\[
	\textrm{BLAT Identity} = \frac{\textrm{Match Length}}{\textrm{Sequence Length}}
\]
Here, the match length refers to the total number of aligned bases that match the reference genome, and the sequence length is the total length of the query sequence.

% Sequences longer than 20 bp were aligned to the reference genome using \gls{blat}.
% The resulting identity scores were used to distinguish between true biological sequences and potential chimeric reads, with lower scores indicating likely artificial sequences.
% To evaluate terminal artificial sequences, we employed the \gls{blat} identity threshold of 0.9 to classify hits and non-hits.
% To assess DeepChopper's accuracy in detecting terminal artificial sequences, we compared its predictions to the soft-clipped regions in untrimmed Dorado-processed data.
% An overlap threshold of 0.4 was used to determine the concordance between the two methods.

\subsection{Heatmap Analysis for chromosomal connections}

The dataset consisted of numerical values associated with 25 chromosomes (chromosome 1 to chromosome 22, chromosome X, chromosome Y and chromosome M).
Each chromosome was represented by a single numerical value, which means the number of connections.
To standardize the data and facilitate comparisons between chromosomes, every value is divided by the length of the related chromosome.

\subsection{Gene expression analysis of \gls{fp} group of chimeric reads}

To identify the \gls{fp} group of chimeric reads, we began by comparing two datasets: the chimeric reads identified by Dorado and those supported by DeepChopper.
The \gls{fp} group was defined by subtracting the chimeric reads confirmed by DeepChopper from the total chimeric reads identified by Dorado.
Following the identification of the \gls{fp} group, we proceeded to assess gene expression levels specifically associated with these reads using IsoQuant (v3.1.2 )~\cite{prjibelski2023accurate}.


\subsection{\gls{go} analysis}

For each cell line, the top 100 genes with high expressed levels were used for the \gls{go} enrichment analysis using the Database for Annotation, Visualization, and Integrated Discovery (DAVID)~\cite{sherman2022david} to elucidate the biological processes they involved.

\subsection{Gene fusion analysis and visualization}

Gene fusions were identified using Fusionseeker (v1.0.1)~\cite{chen2023gene} in Dorado-trimmed reads, Deepchopper-processed reads and \gls{fp} chimeric reads.
The gene structure was visualized using GSDS (v2.0)~\cite{hu2015gsds}.
Base quality scores were generated using a custom Python script, and ONT ion current signals were visualized with Squigualiser (v0.6.3)~\cite{samarakoon2024interactive}.

Visualization of gene fusion events was performed using chimeraviz~\cite{laagstad2017chimeraviz}, a comprehensive R package designed for the analysis and visualization of gene fusion. 
The package was selected for its robust capabilities in generating informative circular plots, also known as Circos plots, which effectively display complex genomic rearrangements.

\subsection{Computing resource}

All computations were performed on a \gls{hpc} server equipped with a 64-core Intel(R) Xeon(R) Gold 6338 CPU and 256 GB of RAM.
The server also featured two NVIDIA A100 \glspl{gpu}, each with 80 GB of GPU memory, providing substantial computational power for both CPU-intensive tasks and GPU-accelerated deep learning workloads.


% \section{Tables}\label{sec5}

% Tables can be inserted via the normal table and tabular environment. To put
% footnotes inside tables you should use \verb+\footnotetext[]{...}+ tag.
% The footnote appears just below the table itself (refer Tables~\ref{tab1} and \ref{tab2}).
% For the corresponding footnotemark use \verb+\footnotemark[...]+

% \begin{table}[h]
% \caption{Caption text}\label{tab1}%
% \begin{tabular}{@{}llll@{}}
% \toprule
% Column 1 & Column 2  & Column 3 & Column 4\\
% \midrule
% row 1    & data 1   & data 2  & data 3  \\
% row 2    & data 4   & data 5\footnotemark[1]  & data 6  \\
% row 3    & data 7   & data 8  & data 9\footnotemark[2]  \\
% \botrule
% \end{tabular}

% \footnotetext{Source: This is an example of table footnote. This is an example of table footnote.}
% \footnotetext[1]{Example for a first table footnote. This is an example of table footnote.}
% \footnotetext[2]{Example for a second table footnote. This is an example of table footnote.}
% \end{table}

% \begin{table}[h]
% 	\caption{Benchmarking for different models}
% 	\label{tab:bechmark}
% 	\begin{tabular}{@{}
% 			l
% 			S[table-format=1.4e-2] % Formats the F1 column in scientific notation
% 			S[table-format=1.4e-2] % Formats the Loss column in scientific notation
% 			@{}}
% 		\toprule
% 		{Model}             & {F1}                         & {Loss}                         \\ \midrule
% 		CNN                 & 0.9909037351608276           & 0.00302332011051476            \\
% 		Model with Hyena    & 0.9925663471221924           & 0.002182388212531805           \\
% 		Model with Caduceus & \bfseries 0.9977396130561829 & \bfseries 0.000541145505849272 \\ \bottomrule
% 	\end{tabular}
% 	% \footnotetext{Source: This is an example of table footnote. This is an example of table footnote.}
% 	% \footnotetext[1]{Example for a first table footnote. This is an example of table footnote.}
% 	% \footnotetext[2]{Example for a second table footnote. This is an example of table footnote.}
% \end{table}


\bmhead{Data Availability}

Raw and processed data from the VCaP cell line supporting the findings of this study have been deposited in the Gene Expression Omnibus (GEO) database under accession code GSE277934.
This dataset includes nanopore \gls{drs} data and nanopore cDNA data.
The deposited data are available in their raw format and in processed forms, including BAM files.
All data are available for public access and can be downloaded directly from the GEO repository (https://www.ncbi.nlm.nih.gov/geo/) using the provided accession number.
The human reference genome GRCh38 was downloaded from \url{http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38\_reference\_genome/}.


\bmhead{Code Availability}

DeepChopper (v1.0) is available on GitHub (\url{https://github.com/ylab-hi/DeepChopper}).
The scripts for model training, performance assessment, and data generation simulation are available on GitHub (\url{https://github.com/ylab-hi/DeepChopper}).
The repository is available under an Apache License, Version 2.0.

\bmhead{Acknowledgements}

Acknowledgments are not compulsory. Where included, they should be brief. Grant or contribution numbers may be acknowledged.
Please refer to the journal-level guidance for any specific requirements.


\backmatter

\begin{appendices}
	\printglossaries[type=\acronymtype, title=Abbreviations]
\end{appendices}


\bibliography{sn-bibliography}% common bib file

\newpage

\section{Extend data}

\renewcommand{\figurename}{Extended Data Fig.}


\begin{figure}[!h]
	\includegraphics[height=0.29\columnwidth]{finals/sf1}
	\caption{ {\bf Analysis of chimeric reads and artificial sequences in VCaP.} (a) Accuracy evaluation in held-out test data set. (b) Chimeric read detection of DeepChopper compared to Dorado (with and without trimming) in the VCaP 004 dataset. The grey bars represent chimeric reads unsupported by \gls{ont} \gls{pcr} cDNA, while the red bars indicate chimeric reads supported by \gls{ont} \gls{pcr} cDNA.  (c) Base quality score and \gls{blat} identity of VCaP RNA004 internal artificial sequences.}\label{fig:sf1}
\end{figure}

\begin{figure}[!h]
	\includegraphics[height=0.26\columnwidth]{finals/sf2}
	\caption{ {\bf Characterization of \gls{fp} chimeric reads Detected by DeepChopper.} (a) Chromosomal distribution and connections in \gls{fp} chimeric reads of VCaP 004. Top: Bar plot showing the number of connections per base pair for each chromosome. Higher bars indicate more frequent connections. Bottom: Heatmap illustrating the count of connections between chromosomes. Intensity of color corresponds to the number of connections. (b) Comparison of gene expression levels between reference genes and \gls{fp} genes in the VCaP 002 dataset. \gls{fp} genes show significantly higher expression levels (\(\textrm{p-value} < 2.2 \times 10^{-16}\)). (c) Distribution of gene effective sizes for all genes and target genes, showing similar lengths between reference and \gls{fp} genes.}\label{fig:sf2}
\end{figure}



\begin{figure}[!h]
	\includegraphics[height=1.6\columnwidth]{finals/sf3}
	\caption{ {\bf Analysis of artificial chimeric reads.} (a) Chromosomal connections of fusion events in false positive chimeric reads from VCaP 002 dataset. Red and blue links indicate intra and inter-chromosomal fusions. (b) \gls{go} enrichment of \gls{fp} chimeric reads detected by DeepChopper in HCT116, A54 and HepG2.}
	\label{fig:sf3}
\end{figure}

% \begin{figure}[!h]
% 	\includegraphics[height=0.87\columnwidth]{finals/sf4}
% 	\caption{ {\bf  Analysis of \gls{fp} chimeric reads } (a) Distribution of \gls{fp} biological processes enriched among \gls{fp} group genes in HCT116. (b) Distribution of \gls{fp} biological processes enriched among \gls{fp} group genes in MCF7. (c) Distribution of \gls{fp} biological processes enriched among \gls{fp} group genes in K562. (d) Distribution of \gls{fp} biological processes enriched among \gls{fp} group genes in HepG2. (e) Distribution of \gls{fp} biological processes enriched among \gls{fp} group genes in VCaP. (f) Distribution of \gls{fp} biological processes enriched among \gls{fp} group genes in A549.}
% 	\label{fig:sf4}
% \end{figure}



\newpage

\section{Supplementary information}

\renewcommand{\figurename}{Supplementary Fig.}
\renewcommand{\tablename}{Supplementary Table}


% If your article has accompanying supplementary file/s please state so here.
% Authors reporting data from electrophoretic gels and blots should supply the full unprocessed scans for key as part of their Supplementary information. This may be requested by the editorial team/s if it is missing.
% Please refer to Journal-level guidance for any specific requirements.


\begin{table}
	\centering
	\caption{Hyperparameter ranges used}\label{tab:hyperparameter}
	\begin{tabular}{lc}
		\toprule
		                     & {$\sf DeepChopper$}               \\
		\midrule
		Parameters           & 1.6M                              \\
		Optimizer            & Adam                             \\
		Optimizer momentum   & $\beta_1$, $\beta_2$ = 0.9, 0.999 \\
		Training epoch       & 60                                \\
		Batch size           & 64                                \\
		Learning rate        & 2e-5                     \\
		LR scheduler         & ReduceLROnPlateau                 \\
		Weight decay (model) & 0                         \\
		Sequence lengths     & 32769                             \\
		\midrule
	\end{tabular}
\end{table}

% \section*{Declarations}

% Some journals require declarations to be submitted in a standardised format. Please check the Instructions for Authors of the journal to which you are submitting to see if you need to complete this section. If yes, your manuscript must contain the following sections under the heading `Declarations':
%
%\begin{itemize}
%	\item Funding
%	\item Conflict of interest/Competing interests (check journal-specific guidelines for which heading to use)
%	\item Ethics approval and consent to participate
%	\item Consent for publication
%	\item Data availability
%	\item Materials availability
%	\item Code availability
%	\item Author contribution
%\end{itemize}

%
%\begin{appendices}
%	\printglossary[type=\acronymtype, title=Abbreviations]
%
%	\section{Section title of first appendix}\label{secA1}
%
%	An appendix contains supplementary information that is not an essential part of the text itself but which may be helpful in providing a more comprehensive understanding of the research problem or it is information that is too cumbersome to be included in the body of the paper.
%
%	%%=============================================%%
%	%% For submissions to Nature Portfolio Journals%%
%	%% please use the heading ``Extended Data''.   %%
%	%%=============================================%%
%
%	%%=============================================================%%
%	%% Sample for another appendix section			               %%
%	%%=============================================================%%
%
%	%% \section{Example of another appendix section}\label{secA2}%
%	%% Appendices may be used for helpful, supporting or essential material that would otherwise
%	%% clutter, break up or be distracting to the text. Appendices can consist of sections, figures,
%	%% tables and equations etc.


%\end{appendices}

%%===========================================================================================%%
%% If you are submitting to one of the Nature Portfolio journals, using the eJP submission   %%
%% system, please include the references within the manuscript file itself. You may do this  %%
%% by copying the reference list from your .bbl file, paste it into the main manuscript .tex %%
%% file, and delete the associated \verb+\bibliography+ commands.                            %%
%%===========================================================================================%%


%% if required, the content of .bbl file can be included here once bbl is generated
%%\input sn-article.bbl

\end{document}
