% !TeX TS-program = lualatex
\documentclass[tikz,border=10pt]{standalone}

\usepackage{fontspec}  % allows for system font usage
\setmainfont{Arial}    % Sets the main font to Arial

\usepackage{xcolor}
\usetikzlibrary{calc}
\usetikzlibrary{positioning}
\usetikzlibrary{arrows.meta}
\usetikzlibrary{decorations.pathreplacing}
\usetikzlibrary{matrix}
\usetikzlibrary{graphs}
\usetikzlibrary{graphs.standard}
\usetikzlibrary{quotes}
\usetikzlibrary{fit}
\usetikzlibrary{chains}
\usetikzlibrary{shapes.geometric}
\usetikzlibrary{shapes.multipart}
\usetikzlibrary{backgrounds}

\usepackage{listofitems} % for \readlist to create arrays

\pgfdeclarelayer{middle}
\pgfsetlayers{background,middle, main}

\begin{document}

\definecolor{c1}{RGB}{239,243,255}
\definecolor{c2}{RGB}{198,219,239}
\definecolor{c3}{RGB}{158,202,225}
\definecolor{c4}{RGB}{107,174,214}
\definecolor{c5}{RGB}{66,146,198}
\definecolor{c6}{RGB}{33,113,181}
\definecolor{c7}{RGB}{8,69,148}


\definecolor{p1}{RGB}{254,224,210}
\definecolor{p2}{RGB}{252,146,114}
\definecolor{p3}{RGB}{222,45,38}

\definecolor{l1}{RGB}{117,107,177}

\colorlet{myred}{red!80!black}
\colorlet{myblue}{blue!80!black}
\colorlet{mygreen}{green!60!black}

% We propose a deep learning model for predicting exitron existence from ultra-long DNA sequences by using a Long-Context Genomic Language Model (GLM) to extract features. The methodology is outlined in several key steps (Figure 1A).

% The input consists of ultra-long sequences, which may include exons and exitrons, with specific nucleotide patterns marked in Figure 1A. During the training phase, sequences are encoded with labels, where exitron sequences are labeled as 1 and others as 0. Sequencing data is processed by a tokenizer, which converts the sequence data into tokens. These tokens are represented by a series of colored blocks in Figure 1A. The tokenized data is then input into a long-context GLM, a model designed to handle long-range dependencies and consider distant relationships within the sequence (Figure 1A).

% The GLM can extract concise features and help us understand long sequences better. These extracted features are subsequently fed into a Multi-Layer Perceptron (MLP). The processed features from the MLP are directed to the output layer, where the final predictions are made. The output is compared with the original sequencing labels to compute the loss and update the model parameters accordingly. In conclusion, the model extracts sequencing features using a sophisticated long-context GLM as well as effectively analyzing ultra-long DNA sequences to predict exitron existence Â (Figure 1A).


\begin{tikzpicture}[
	font=\scriptsize,
	block/.style={rectangle, text centered, rounded corners, minimum height=1em, minimum width=4em, fill=c6},
	connect/.style={rounded corners, black!30, text=black, very thick, -{Stealth[length=5pt, width=4pt]}, shorten >=1pt, shorten <=1pt},
	skip loop/.style  2 args= {connect, to path={-- ++(#1, 0) |- node[right, sloped, pos=0.2, inner sep=1pt, fill=white, font=\small] {#2} (\tikztotarget)}},
	nnode/.style={thin, circle,draw=c7, minimum size=6, inner sep=0.5,outer sep=0.6},
	node in/.style={nnode, green!20!black,draw=c7!30!black,fill=c7!25},
	node hidden/.style={
			nnode,
			blue!20!black,
			draw=myblue!30!black,
			fill=myblue!20,
		},
	node out/.style={nnode, red!20!black, draw=myred!30!black,fill=myred!20},
	]

	% Exitron and Exon
	\begin{scope}[local bounding box=exon, scale=1.4, transform shape,
		]
		\begin{scope}[start chain, node distance=0mm,
				nodes/.style={inner sep=0, outer sep=0},
			]
			\node (exon-n1) [on chain] {ATT};
			\node (exon-n2) [fill=c7, text=white, on chain] {GT$\cdots$AG};
			\node (exon-n3) [on chain] {TCGG};
			\node (exon-n4) [fit=(exon-n1)(exon-n2)(exon-n3), draw=c7,  thick, inner sep=0] {};
			\node (exitron-label) [above=1pt of exon-n4] {Exitron};
			\node (exitron-label2) [above right= of exon-n4, xshift=-7] {Exon};
		\end{scope}

		\node (left) [left= 20pt of exon-n1]  {};
		\node (right) [right= 10pt of exon-n3]   {};

		\path (left) edge[-, thick]  (exon-n1)
		(exon-n3) edge[-, thick] (right);

		\node (dnalabel) [above=8pt of exon-n1, xshift=-6pt] {Ultra-long DNA};
	\end{scope}


	% RNAseq Matrix
	\begin{scope}[shift={(exon.center)},
			yshift=-1.9cm, xshift=14,
			node distance=0mm, scale=1, transform shape,
			every node/.style={scale=1, transform shape},
		]
		\matrix (rnaseq) [matrix of nodes, nodes in empty cells,
			column sep=-\pgflinewidth, row sep=1pt,
			nodes={thin, inner sep=0pt, outer sep=0pt,
					minimum width=1.3em,
					minimum height=1.3em,
					anchor=center},
			row 1/.style={nodes={fill=p1}},
			row 1 column 4/.style={nodes={fill=p3}},
			row 1 column 5/.style={nodes={fill=p3}},
			row 1 column 6/.style={nodes={fill=p3}},
			row 1 column 7/.style={nodes={fill=p3}},
			row 1 column 8/.style={nodes={fill=p3}},
		] {
			0 & 0 & 0 & 1 & 1 & $\cdots$ & 1 & 1 & 0 & 0 & 0 & 0 \\
			A & T & T & G & T & $\cdots$ & A & G & T & C & G & G \\
		};
		\node (rnaseq-label) [left=1pt of rnaseq-2-1, font=\small] {DNA};
		\node (label) [left=1pt of rnaseq-1-1, font=\small] {Label};
		\node (databox) [fit=(rnaseq)(rnaseq-label)(label), draw=c7, dashed, thick, inner sep=0] {};
	\end{scope}


	\begin{scope}[local bounding box=tokens, shift={(databox.south)}, yshift=-0.65cm,
			every node/.style={scale=1.3, transform shape},
		]
		\matrix (tokens1)   [
			matrix of nodes,
			nodes in empty cells,
			nodes={fill=c2, draw=black, thin, outer sep=0pt},
			column sep=-\pgflinewidth, row sep=-\pgflinewidth,
			column 1/.style={nodes={fill=c1}},
			column 2/.style={nodes={fill=c2}},
			column 3/.style={nodes={fill=c3}},
			column 4/.style={nodes={fill=c4}},
			column 5/.style={nodes={fill=c5}},
			column 6/.style={nodes={fill=c6}},
			column 7/.style={nodes={fill=c7}},
		]
		{
			 &  &  &  &  &  & \\
		};
	\end{scope}

	\begin{scope}[shift={(tokens1)}, yshift=-1.8cm, xshift=-2.87cm,
		edge1/.style={black, line width=0.6pt, -{Stealth[length=3pt, width=2pt]}, shorten >=0.5pt, shorten <=0.5pt},
		local bounding box=lcglm, scale=0.95, transform shape]
		\graph [
			grow right = 1cm,
			edges = {edge1},
			skip loop/.style={to path={ -- ++(1cm, 0) -| (\tikztotarget) \tikztonodes}},
			hyena/.style ={rectangle, draw=c7, fill=c7!20, thick,rotate=270,  minimum height=2em, minimum width=5em, font=\large},
			residual/.style={circle, draw=c7, line width=0.6pt, inner sep=1, as=+},
			block1/.style={rectangle, draw=c5, fill=c5!70, thick,   minimum height=3em, minimum width=1em, as=},
			block2/.style={rectangle, draw=c6, fill=c6!70, thick,  minimum height=4em, minimum width=1em, as=},
		] {
			input [as=] -> hyena [hyena] -> r1 [residual] -> b1 [block1] -> b2 [block2] -> r2 [residual] -> b3 [block1];
		};

		\draw[edge1]  ($(input) !.2! (hyena)$) |- ++(0,1) -| (r1);
		\draw[edge1] ($(b1) !.4! (b2) $) |- ++(0,1) -| (r2);
	\end{scope}


	\begin{pgfonlayer}{middle}
		\node (lcglmbox) [fit=(lcglm), fill=l1,  thick, inner ysep=6pt, inner xsep=2pt, yshift=5pt, opacity=0.7] {};
		\node[inner sep=3pt, anchor=north west] at (lcglmbox.north west) {Long-Context GLM};
	\end{pgfonlayer}


	\begin{scope}[shift={(lcglmbox.south)}, yshift=-0.5cm, xshift=3, local bounding box=features,
			every node/.style={scale=1.3, transform shape},
		]
		\begin{scope}[opacity=0.8]
			\matrix (features1)   [
				matrix of nodes,
				nodes in empty cells,
				nodes={fill=c2,draw},
				column sep=-\pgflinewidth, row sep=-\pgflinewidth,
				column 1/.style={nodes={fill=c1}},
				column 2/.style={nodes={fill=c2}},
				column 3/.style={nodes={fill=c3}},
				column 4/.style={nodes={fill=c4}},
				column 5/.style={nodes={fill=c5}},
				column 6/.style={nodes={fill=c6}},
				column 7/.style={nodes={fill=c7}},
			]
			{
				 &  &  &  &  &  & \\
			};
		\end{scope}

		\begin{scope}[shift={(-0.1, -0.1)}, opacity=0.9]
			\matrix (features2)   [
				matrix of nodes,
				nodes in empty cells,
				nodes={fill=c2, draw, outer sep=0pt},
				column sep=-\pgflinewidth, row sep=-\pgflinewidth,
				column 1/.style={nodes={fill=c1}},
				column 2/.style={nodes={fill=c2}},
				column 3/.style={nodes={fill=c3}},
				column 4/.style={nodes={fill=c4}},
				column 5/.style={nodes={fill=c5}},
				column 6/.style={nodes={fill=c6}},
				column 7/.style={nodes={fill=c7}},
			]
			{
				 &  &  &  &  &  & \\
			};
		\end{scope}

		\begin{scope}[shift={(-0.2, -0.2)}, opacity=1]
			\matrix (features3)   [
				matrix of nodes,
				nodes in empty cells,
				nodes={fill=c2, draw, outer sep=0pt},
				column sep=-\pgflinewidth, row sep=-\pgflinewidth,
				column 1/.style={nodes={fill=c1}},
				column 2/.style={nodes={fill=c2}},
				column 3/.style={nodes={fill=c3}},
				column 4/.style={nodes={fill=c4}},
				column 5/.style={nodes={fill=c5}},
				column 6/.style={nodes={fill=c6}},
				column 7/.style={nodes={fill=c7}},
			]
			{
				 &  &  &  &  &  & \\
			};
		\end{scope}

	\end{scope}

	\begin{scope}[shift={(lcglm)}, yshift=-1.6cm, x=2.2cm, y=1.4cm, local bounding box=mlp,
			mynode/.style={thick, draw=c7,fill=c7!20, circle, minimum size=18},
			rotate=-90,
			scale=0.4,
			transform shape,
		]
		\readlist\Nnod{3,4,2} % number of nodes per layer
		% \Nnodlen = length of \Nnod (i.e. total number of layers)
		% \Nnod[1] = element (number of nodes) at index 1
		\foreachitem \N \in \Nnod{ % loop over layers
		% \N     = current element in this iteration (i.e. number of nodes for this layer)
		% \Ncnt  = index of current layer in this iteration
		\foreach \i [evaluate={\x=\Ncnt; \y=\N/2-\i+0.5; \prev=int(\Ncnt-1);}] in {1,...,\N}{ % loop over nodes
		\node[mynode] (N\Ncnt-\i) at (\x,\y) {};
		\ifnum\Ncnt>1 % connect to previous layer
			\foreach \j in {1,...,\Nnod[\prev]}{ % loop over nodes in previous layer
			\draw[draw=black!80,  line width=0.6pt, -{Stealth[length=3pt, width=2pt]},
			shorten >=0.5pt, shorten <=0.5pt] (N\prev-\j) -- (N\Ncnt-\i); % connect arrows directly
			}
		\fi % else: nothing to connect first layer
		}
		}
	\end{scope}

	\node [left = of N2-1, xshift=-25pt, yshift=-2pt, font=\small] {MLP};


	\begin{scope}[local bounding box=outputbox, shift={(mlp.south)}, yshift=-0.6cm]
		\matrix (output)   [
			matrix of nodes,
			nodes in empty cells,
			nodes={fill=c2, draw=black, thin, outer sep=0pt, scale=1.3, transform shape},
			column sep=-\pgflinewidth, row sep=-\pgflinewidth,
			column 1/.style={nodes={fill=p1}},
			column 2/.style={nodes={fill=p1}},
			column 3/.style={nodes={fill=p3}},
			column 4/.style={nodes={fill=p3}},
			column 5/.style={nodes={fill=p1}},
			column 6/.style={nodes={fill=p1}},
			column 7/.style={nodes={fill=p1}},
		]
		{
			 &  &  &  &  &  & \\
		};
	\end{scope}

	\node [left = of output, xshift=5pt, anchor=center, font=\small] {Output};

	\begin{scope}[shift={(output)}, yshift=-1cm, local bounding box=prediction, scale=1, transform shape,
			every node/.style={scale=1.3, transform shape},
		]
		\matrix (mprediction) [matrix of nodes, nodes in empty cells,
			column sep=-\pgflinewidth, row sep=-\pgflinewidth,
			nodes={thin,  inner sep=0pt, outer sep=0pt, minimum width=0.6em, minimum height=0.6em, anchor=center},
			row 1 column 4/.style={nodes={fill=p3, text=white}},
			row 1 column 5/.style={nodes={fill=p3, text=white}},
			row 1 column 6/.style={nodes={fill=p3, text=white}},
			row 1 column 7/.style={nodes={fill=p3, text=white}},
			row 1 column 8/.style={nodes={fill=p3, text=white}},
		] { A & T & T & G & T & $\cdots$ & A & G & T & C & G & G \\
		};
	\end{scope}


	\path
	(exon) edge[connect] (databox)
	(databox) edge["Tokenizer", connect] (tokens)
	(mlp) edge[connect] (output)
	(features) edge[connect] (mlp)
	(output) edge["Predict", connect, font=\small] (prediction)
	edge[blue,  skip loop={36mm}{Loss}] (rnaseq-1-12);

	\draw [connect] (tokens1.south) |- ++(-9.0em, -0.25em) |-  ($(input) + (0.1em, 0)$);

	\draw [connect] ($(b3.east) + (1pt, 0)$) -| ++ (0.5em, -2.7em)  -|   ($(features1) + (-0.3em, 0.3em)$);


	\begin{scope}[on background layer]
		\node (modelbox) [fit=(tokens)(lcglm)(features)(mlp), fill=c2,  thick, inner sep=8pt] {};
		\node[inner sep=2pt, anchor=north west, font=\small] at (modelbox.north west) {Model};
	\end{scope}

\end{tikzpicture}

\end{document}
